File: opt/gurobi501/linux64/examples/java/Callback.java
/* Copyright 2012, Gurobi Optimization, Inc. */

/* This example reads an LP or a MIP from a file, sets a callback
   to monitor the optimization progress and to output some progress
   information to the screen and to a log file. If it is a MIP and 10%
   gap is reached, then it aborts */

import gurobi.*;

public class Callback extends GRBCallback {
  private GRBVar[] vars;
  private int      lastmsg;

  public Callback(GRBVar[] xvars) {
    vars = xvars;
    lastmsg = -100;
  }

  protected void callback() {
    try {
      if (where == GRB.CB_MESSAGE) {
        String st = getStringInfo(GRB.CB_MSG_STRING);
        if (st != null) System.out.print(st);
      } else if (where == GRB.CB_PRESOLVE) {
        int cdels = getIntInfo(GRB.CB_PRE_COLDEL);
        int rdels = getIntInfo(GRB.CB_PRE_ROWDEL);
        System.out.println(cdels+" columns and "+rdels+" rows are removed");
      } else if (where == GRB.CB_SIMPLEX) {
        int itcnt = (int) getDoubleInfo(GRB.CB_SPX_ITRCNT);
        if (itcnt%100 == 0) {
          double obj  = getDoubleInfo(GRB.CB_SPX_OBJVAL);
          double pinf = getDoubleInfo(GRB.CB_SPX_PRIMINF);
          double dinf = getDoubleInfo(GRB.CB_SPX_DUALINF);
          int  ispert = getIntInfo(GRB.CB_SPX_ISPERT);
          char ch;
          if (ispert == 0)      ch = ' ';
          else if (ispert == 1) ch = 'S';
          else                  ch = 'P';
          System.out.println(itcnt+"  "+ obj + ch + "  "+pinf + "  " + dinf);
        }
      } else if (where == GRB.CB_MIP) {
        int nodecnt = (int) getDoubleInfo(GRB.CB_MIP_NODCNT);
        if (nodecnt - lastmsg >= 100) {
          lastmsg = nodecnt;
          double objbst = getDoubleInfo(GRB.CB_MIP_OBJBST);
          double objbnd = getDoubleInfo(GRB.CB_MIP_OBJBND);
          if (Math.abs(objbst - objbnd) < 0.1 * (1.0 + Math.abs(objbst)))
            abort();
          int actnodes = (int) getDoubleInfo(GRB.CB_MIP_NODLFT);
          int itcnt    = (int) getDoubleInfo(GRB.CB_MIP_ITRCNT);
          int solcnt   = getIntInfo(GRB.CB_MIP_SOLCNT);
          int cutcnt   = getIntInfo(GRB.CB_MIP_CUTCNT);
          System.out.println(nodecnt + " " +  actnodes + " " +  itcnt + " "
            +  objbst + " " +  objbnd + " " +  solcnt + " " +  cutcnt);
        }
      } else if (where == GRB.CB_MIPSOL) {
        double obj     = getDoubleInfo(GRB.CB_MIPSOL_OBJ);
        int    nodecnt = (int) getDoubleInfo(GRB.CB_MIPSOL_NODCNT);
        double[] x = getSolution(vars);
        System.out.println("**** New solution at node " + nodecnt + ", obj "
                           + obj + ", x[0] = " + x[0] + "****");
      }
    } catch (GRBException e) {
      System.out.println("Error code: " + e.getErrorCode() + ". " +
          e.getMessage());
      e.printStackTrace();
    }
  }

  public static void main(String[] args) {

    if (args.length < 1) {
      System.out.println("Usage: java Callback filename");
      System.exit(1);
    }

    try {
      GRBEnv    env   = new GRBEnv();
      GRBModel  model = new GRBModel(env, args[0]);

      model.getEnv().set(GRB.IntParam.OutputFlag, 0);

      GRBVar[] vars   = model.getVars();

      model.setCallback(new Callback(vars));
      model.optimize();

      double[] x      = model.get(GRB.DoubleAttr.X, vars);
      String[] vnames = model.get(GRB.StringAttr.VarName, vars);

      for (int j = 0; j < vars.length; j++) {
        if (x[j] != 0.0) System.out.println(vnames[j] + " " + x[j]);
      }

      for (int j = 0; j < vars.length; j++) {
        if (x[j] != 0.0) System.out.println(vnames[j] + " " + x[j]);
      }

      // Dispose of model and environment
      model.dispose();
      env.dispose();

    } catch (GRBException e) {
      System.out.println("Error code: " + e.getErrorCode() + ". " +
          e.getMessage());
      e.printStackTrace();
    }
  }
}

File: opt/gurobi501/linux64/examples/python/callback.py
#!/usr/bin/python

# Copyright 2012, Gurobi Optimization, Inc.

# This example reads an LP or a MIP from a file, sets a callback
# to monitor the optimization progress, and outputs progress
# information to the screen and to the log file. If the input model
# is a MIP, the callback aborts the optimization after 10000 nodes have
# been explored.


import sys
from gurobipy import *


# Define callback function

def mycallback(model, where):
    if where == GRB.callback.PRESOLVE:
        print 'Removed', model.cbGet(GRB.callback.PRE_COLDEL), 'columns and', \
               model.cbGet(GRB.callback.PRE_ROWDEL), 'rows'
    elif where == GRB.callback.SIMPLEX:
        itcnt = model.cbGet(GRB.callback.SPX_ITRCNT)
        if itcnt % 100 == 0:
            obj  = model.cbGet(GRB.callback.SPX_OBJVAL)
            pinf = model.cbGet(GRB.callback.SPX_PRIMINF)
            dinf = model.cbGet(GRB.callback.SPX_DUALINF)
            pert = model.cbGet(GRB.callback.SPX_ISPERT)
            if pert == 0:
                ch = ' '
            elif pert == 1:
                ch = 'S'
            else:
                ch = 'P'
            print int(itcnt), obj, ch, pinf, dinf
    elif where == GRB.callback.MIP:
        nodecnt = model.cbGet(GRB.callback.MIP_NODCNT)
        if nodecnt % 100 == 0:
            objbst = model.cbGet(GRB.callback.MIP_OBJBST)
            objbnd = model.cbGet(GRB.callback.MIP_OBJBND)
            print int(nodecnt), objbst, objbnd
        if nodecnt > model._mynodelimit:
            model.terminate()
    elif where == GRB.callback.MIPSOL:
        obj     = model.cbGet(GRB.callback.MIPSOL_OBJ)
        nodecnt = int(model.cbGet(GRB.callback.MIPSOL_NODCNT))
        print '*** New solution at node', nodecnt, 'objective', obj
        print model.cbGetSolution(model.getVars())
    elif where == GRB.callback.MIPNODE:
        print '*** New node'
        if model.cbGet(GRB.callback.MIPNODE_STATUS) == GRB.status.OPTIMAL:
            x = model.cbGetNodeRel(model.getVars())
            model.cbSetSolution(model.getVars(), x)

if len(sys.argv) < 2:
    print 'Usage: callback.py filename'
    quit()

# Read and solve model

model = read(sys.argv[1])

# Pass data into my callback function

model._mynodelimit = 10000

model.params.heuristics = 0

model.optimize(mycallback)
